compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1577,6 +1589,13 @@ impl Type {
         errors: &mut Vec<TypeCheckError>,
         make_error: impl FnOnce() -> TypeCheckError,
     ) {
+        // Try to coerce `fn (..) -> T` to `unconstrained fn (..) -> T`
+        let Some(make_error) = self.try_fn_to_unconstrained_fn_coercion(
+            expected, expression, interner, errors, make_error,
+        ) else {
+            return;
+        };
```

Comment:
This should be only after we fail `try_unify`, like `try_array_to_slice_coercion` is. We can combine it with that check and get rid of the early return.

Reply:
Done! I did it one way first, then I wasn't sure it was clear with all that nesting so... now it's doing early return again (many times) but maybe it's clearer this time.